### ISSUE TITLE
pl.read_csv(gzipped, n_rows=2) is 10,000 times slower than with n_rows=1

### ISSUE THREAD

**Issue by Wainberg** [pola-rs/polars#10579]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
from subprocess import run
from timeit import timeit

run('wget https://figshare.com/ndownloader/files/34517861 -O PGC3_SCZ_wave3.primary.autosome.public.v3.vcf.tsv.gz', shell=True)
print(timeit(lambda: pl.read_csv('PGC3_SCZ_wave3.primary.autosome.public.v3.vcf.tsv.gz', n_rows=1), number=1))  # 0.000583793967962265
print(timeit(lambda: pl.read_csv('PGC3_SCZ_wave3.primary.autosome.public.v3.vcf.tsv.gz', n_rows=2), number=1))  # 5.308462973218411
print(timeit(lambda: pl.read_csv('PGC3_SCZ_wave3.primary.autosome.public.v3.vcf.tsv.gz'), number=1))  # 5.346831194125116
```

### Issue description

[This](https://figshare.com/ndownloader/files/34517861) 225 MB gzipped TSV (with multiline header) contains the main results from the largest-ever genetic study of schizophrenia, and was provided by the researchers themselves. (I'm aware that this is a suboptimal way to store this much tabular data, but it's the standard in the field.) After downloading it, reading one row is very fast, but reading two rows is about as slow as reading the whole file. 

Notably, the critical number of rows depends on the separator. With `separator=' '`, reading 13 rows is very fast but 14 rows takes as long as reading the entire file. With `separator='\t'`, reading 70 rows is fast but 71 isn't. However, for a given separator, the critical number of rows is the same with `has_header=True` and `has_header=False`.

### Expected behavior

Reading N+1 rows should be about as fast as reading N rows. 

### Installed versions

<details>

```
--------Version info---------
Polars:              0.18.15
Index type:          UInt32
Platform:            Linux-3.10.0-1160.88.1.el7.x86_64-x86_64-with-glibc2.17
Python:              3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:17) [GCC 12.2.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
matplotlib:          3.7.2
numpy:               1.25.2
pandas:              2.0.3
pyarrow:             12.0.1
pydantic:            <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            0.8.1
xlsxwriter:          3.1.2
```

</details>

**Comment by mcrumiller** [comment]
Hi @Wainberg,

Did you look at the file you're trying to read? There are 73 lines of comments before the data begins. Can you try again with `skip_rows=73`? If you still have issues after that, we can try to diagnose. Most likely, `read_csv` thinks your first line is the header, and then has trouble deciphering after that, which is very reasonable.

**Comment by Wainberg** [comment]
That's exactly right - like I said, it's got a multiline header. The problem is still there with `skip_rows=73` though, and what's funny is that it occurs after the same number of lines (70 vs 71) as when `skip_rows` wasn't specified. The dataframe looks completely normal in both cases, it just takes 1000x longer to load the one extra line.

```python
>>> timeit(lambda: pl.read_csv('PGC3_SCZ_wave3.primary.autosome.public.v3.vcf.tsv.gz', separator='\t', skip_rows=73, n_rows=70), number=1)
0.0014153728261590004
>>> timeit(lambda: pl.read_csv('PGC3_SCZ_wave3.primary.autosome.public.v3.vcf.tsv.gz', separator='\t', skip_rows=73, n_rows=71), number=1)
4.5186741831712425
```

**Comment by mcrumiller** [comment]
@Wainberg, your file definitely includes some non-standard formatting. When I deleted the header columns and ran a standard `read_csv`, I encountered this:

![image](https://github.com/pola-rs/polars/assets/1896992/97cd1ca9-8064-4ad1-bd4e-393a8fb1010c)

There are lines in the file that cut short. This will probably bork our CSV reader. This is much farther into the file than you described, do you want to still investigate the issue in earlier rows?

**Comment by Wainberg** [comment]
It's malformed for sure, but I was able to reproduce the "wrong number of rows" bug with a more standard file, so I made a [separate bug report](https://github.com/pola-rs/polars/issues/10584) for it.

**Comment by Wainberg** [comment]
Ok, I was just able to reproduce the other issue with that file as well - try downloading `https://raw.githubusercontent.com/hadley/mastering-shiny/main/neiss/injuries.tsv.gz` and running these:

```python
>>> timeit(lambda: pl.read_csv('injuries.tsv.gz', separator='@', n_rows=124), number=1)
0.00462570006493479
>>> timeit(lambda: pl.read_csv('injuries.tsv.gz', separator='@', n_rows=125), number=1)
0.6000783999916166
```
The first `@` appears on line 126 (if the line after the header is line 0), so I'm guessing that when it gets to `n_rows=125` it decompresses a chunk of the gzip file that contains the first `@` (which is two lines past the end of what it's supposed to read), realizes that the file is ragged (variable number of columns per row), and then for some reason switches to a 100x slower approach. 

In addition to being slower, it's also incorrect - it silently discards everything after the `@` without giving an error. I've filed a separate bug report about that [here](https://github.com/pola-rs/polars/issues/10585) because it happens with non-gzipped files as well. Hopefully fixing that issue will also fix the large runtime difference.

**Comment by mcrumiller** [comment]
In your most recent file, the appropriate separator is `\t`, not `@`--the `@` is just a character in some of the strings. If you use `\t` it reads it in just fine into columns and I don't see the row problem.

I can reproduce the odd timing, but it doesn't seem to be very sensible to diagnose this since `@` is not the correct separator--when you use `@`, you hit a newline prior to the `@` symbol, and so polars infers (correctly) that there is only one column, and your entire TSV file is read in with one column:

```python
import polars as pl

s = pl.read_csv("injuries.tsv.gz", separator="\t", n_rows=124)
s.head()
```
```
shape: (5, 10)
┌────────────┬─────┬──────┬───────┬───┬───────────────────────┬───────────┬─────────┬───────────────────────────────────┐
│ trmt_date  ┆ age ┆ sex  ┆ race  ┆ … ┆ location              ┆ prod_code ┆ weight  ┆ narrative                         │
│ ---        ┆ --- ┆ ---  ┆ ---   ┆   ┆ ---                   ┆ ---       ┆ ---     ┆ ---                               │
│ str        ┆ i64 ┆ str  ┆ str   ┆   ┆ str                   ┆ i64       ┆ f64     ┆ str                               │
╞════════════╪═════╪══════╪═══════╪═══╪═══════════════════════╪═══════════╪═════════╪═══════════════════════════════════╡
│ 2017-01-01 ┆ 71  ┆ male ┆ white ┆ … ┆ Other Public Property ┆ 1807      ┆ 77.6641 ┆ 71YOM FELL ON FLOOR IN BATHROOM.… │
│ 2017-01-01 ┆ 16  ┆ male ┆ white ┆ … ┆ Home                  ┆ 676       ┆ 77.6641 ┆ 16YOM TRIPPED OVER A RUG AND FEL… │
│ 2017-01-01 ┆ 58  ┆ male ┆ white ┆ … ┆ Home                  ┆ 649       ┆ 77.6641 ┆ 58 YOM WENT TO BATHROOM;  LOST H… │
│ 2017-01-01 ┆ 21  ┆ male ┆ white ┆ … ┆ Home                  ┆ 4076      ┆ 77.6641 ┆ 21 YOM TURNED WRONG AS HE GOT OU… │
│ 2017-01-01 ┆ 54  ┆ male ┆ white ┆ … ┆ Other Public Property ┆ 1807      ┆ 77.6641 ┆ 54 YOM WAS FOUND ON FLOOR OF NUR… │
└────────────┴─────┴──────┴───────┴───┴───────────────────────┴───────────┴─────────┴───────────────────────────────────┘
```
versus:
```python
s = pl.read_csv("injuries.tsv.gz", separator="@", n_rows=124)
s.head()
```
```
shape: (5, 1)
┌───────────────────────────────┐
│ trmt_date     age     sex     race    body_part… │
│ ---                           │
│ str                           │
╞═══════════════════════════════╡
│ 2017-01-01    71      male    white   Upper T… │
│ 2017-01-01    16      male    white   Lower A… │
│ 2017-01-01    58      male    white   Upper T… │
│ 2017-01-01    21      male    white   Lower T… │
│ 2017-01-01    54      male    white   Head    In…  │
└───────────────────────────────┘
```

I'm not sure why 125 rows takes much longer than 124 rows when you read in each row into a single column, but it's most likely triggering a large preallocation. Do you still want to look into this?

**Comment by Wainberg** [comment]
I mean this is a totally contrived example, and it's fine if you want to close it since it should be resolved once #10585 is dealt with.

**Comment by Wainberg** [comment]
Closing since the resolution of #10585 makes these examples give an error by default anyway.
